### PR TITLE
Cloud ObjectStore

### PR DESCRIPTION
**Pull request by VJalili**
The `Cloud` ObjectStore is a provider-agnostic cloud-based storage. In other words, the `Cloud` ObjectStore can read/write from various cloud-based storages without a provider-specific implementation; it leverages on [`CloudBridge`](https://github.com/gvlproject/cloudbridge) which provides a unified access to various providers. 

This PR sets a base for the User-based ObjectStore we have been working on. Initially, the `Cloud` ObjectStore was part of the User-based ObjectStore (PR [#4314](https://github.com/galaxyproject/galaxy/pull/4314)), and it is being submitted as a separate PR following the comments on the PR [#4314](https://github.com/galaxyproject/galaxy/pull/4314).

To simplify the comparison with `S3ObjectStore`, the `Cloud` ObjectStore provides the very same functionality as `S3ObjectStore`, i.e., read/write to AWS S3 buckets. The only difference is: `S3ObjectStore` leverages on `boto` while `Cloud` uses `CloudBridge`. 

pinging @bgruening @nsoranzo and @erasche

### CODE REVIEW

File: lib/galaxy/objectstore/cloud.py [hunk 1]
```diff
@@ -0,0 +1,615 @@
...
+
+from ..objectstore import convert_bytes, ObjectStore
+from cloudbridge.cloud.factory import CloudProviderFactory, ProviderList
+
+try:
+    # Imports are done this way to allow objectstore code to be used outside of Galaxy.
+    import boto
+
+    from boto.exception import S3ResponseError
+    from boto.s3.key import Key
```
Comment (mvdbeek) [inline lib/galaxy/objectstore/cloud.py]: The linter doesn't like unused imports (this line and the following), but you can ignore this by doing
```
from boto.s3.key import Key  # noqa: F401
from boto.s3.connection import S3Connection  # noqa: F401
```

Although I'm not sure in which scenario it would help to import unused dependencies ...
If external code needs it, it should probably import the dependencies where they are needed.

Comment (VJalili) [reply lib/galaxy/objectstore/cloud.py]: @mvdbeek Thanks, the imports are updated accordingly. 
Actually, `boto` import is temporary solution for handling AWS S3 exceptions until `cloudbridge` implements such exceptions for all the providers.

File: lib/galaxy/dependencies/conditional-requirements.txt [hunk 2]
```diff
@@ -13,6 +13,7 @@ graphitesend
 azure-storage==0.32.0
 # PyRods not in PyPI
 python-ldap==2.4.27
+cloudbridge==0.3.1
```
Comment (afgane) [inline lib/galaxy/dependencies/conditional-requirements.txt]: Any reason not to go with the latest current release (0.3.2)?

File: lib/galaxy/objectstore/cloud.py [hunk 3]
```diff
@@ -0,0 +1,611 @@
...
+    safe_relpath,
+    string_as_bool,
+    umask_fix_perms,
+)
+from galaxy.util.sleeper import Sleeper
+
+from ..objectstore import convert_bytes, ObjectStore
+from cloudbridge.cloud.factory import CloudProviderFactory, ProviderList
+
+try:
```
Comment (afgane) [inline lib/galaxy/objectstore/cloud.py]: Presumably, this is for backward compatibility? Either way, it'd be good to have a comment why is this necessary in addition to the CloudBridge import.

File: lib/galaxy/objectstore/cloud.py [hunk 4]
```diff
@@ -0,0 +1,611 @@
...
+        it a few times. Raise error if connection is not established. """
+        for i in range(5):
+            try:
+                bucket = self.conn.object_store.get(bucket_name)
+                if bucket is None:
+                    log.debug("Bucket not found, creating a bucket with handle '%s'", bucket_name)
+                    bucket = self.conn.object_store.create(bucket_name)
+                log.debug("Using cloud object store with bucket '%s'", bucket.name)
+                return bucket
+            except S3ResponseError:
```
Comment (afgane) [inline lib/galaxy/objectstore/cloud.py]: This is not imported anywhere. Also, `create` method returns `S3CreateError` (http://boto.cloudhackers.com/en/latest/ref/boto.html#module-boto.exception). 
More importantly, this will not work cross-provider so having an extra, general `except` may be desirable?
Same comment applies to other refs to `S3ResponseError`.
(xref https://github.com/gvlproject/cloudbridge/issues/9 and custom/standardized CloudBridge exceptions).

File: lib/galaxy/objectstore/cloud.py [hunk 5]
```diff
@@ -0,0 +1,611 @@
...
+        self.conn = CloudProviderFactory().create_provider(ProviderList.AWS, aws_config)
+
+    def _parse_config_xml(self, config_xml):
+        try:
+            a_xml = config_xml.findall('auth')[0]
+            self.access_key = a_xml.get('access_key')
+            self.secret_key = a_xml.get('secret_key')
+            b_xml = config_xml.findall('bucket')[0]
+            self.bucket = b_xml.get('name')
+            self.use_rr = string_as_bool(b_xml.get('use_reduced_redundancy', "False"))
```
Comment (afgane) [inline lib/galaxy/objectstore/cloud.py]: It doesn't look like we're using this option anymore (which is understandable from the multi-cloud perspective), so probably best to remove this line and relevant config.

File: lib/galaxy/objectstore/cloud.py [hunk 6]
```diff
@@ -0,0 +1,611 @@
...
+                              rel_path)
+                    return True
+                # FIXME: don't need to differenciate between uploading from a string or file,
+                # because CloudBridge handles this internally.
+                if from_string:
+                    if not self.bucket.get(rel_path):
+                        created_obj = self.bucket.create_object(rel_path)
+                        created_obj.upload(source_file)
+                    else:
+                        self.bucket.get(rel_path).upload(source_file)
```
Comment (afgane) [inline lib/galaxy/objectstore/cloud.py]: Should probably keep the same code structure as in the `if` case (or the other way around, as long as it's consistent).

File: lib/galaxy/objectstore/cloud.py [hunk 7]
```diff
@@ -0,0 +1,611 @@
...
+        """
+        try:
+            source_file = source_file if source_file else self._get_cache_path(rel_path)
+            if os.path.exists(source_file):
+                if os.path.getsize(source_file) == 0 and (self.bucket.get(rel_path) is not None):
+                    log.debug("Wanted to push file '%s' to S3 key '%s' but its size is 0; skipping.", source_file,
+                              rel_path)
+                    return True
+                # FIXME: don't need to differenciate between uploading from a string or file,
+                # because CloudBridge handles this internally.
```
Comment (afgane) [inline lib/galaxy/objectstore/cloud.py]: What's there to fix? `if` and `else` here look the same so shouldn't it just be consolidated?

File: lib/galaxy/objectstore/cloud.py [hunk 8]
```diff
@@ -0,0 +1,611 @@
...
+                    rel_path = os.path.join(extra_dir, rel_path)
+                else:
+                    rel_path = os.path.join(rel_path, extra_dir)
+
+            # Create given directory in cache
+            cache_dir = os.path.join(self.staging_path, rel_path)
+            if not os.path.exists(cache_dir):
+                os.makedirs(cache_dir)
+
+            # Although not really necessary to create S3 folders (because S3 has
```
Comment (afgane) [inline lib/galaxy/objectstore/cloud.py]: Should probably remove this comment.

File: lib/galaxy/objectstore/cloud.py [hunk 9]
```diff
@@ -0,0 +1,611 @@
...
+                return True
+
+            # For the case of extra_files, because we don't have a reference to
+            # individual files/keys we need to remove the entire directory structure
+            # with all the files in it. This is easy for the local file system,
+            # but requires iterating through each individual key in S3 and deleing it.
+            if entire_dir and extra_dir:
+                shutil.rmtree(self._get_cache_path(rel_path))
+                results = self.bucket.list(prefix=rel_path)
+                for key in results:
```
Comment (afgane) [inline lib/galaxy/objectstore/cloud.py:461]: May be better to replace references like this that use AWS terminology to something cloud agnostic? e.g., `object`. This applies throughout the file (e.g., `_key_exists`).

**Review by afgane — CHANGES_REQUESTED** [review]
Looks pretty good. Clearly, this is still AWS-only even if it implements the CloudBridge interface but I guess that's just the first step. 
There are a few comment lines and debug print statements that were also in the previous implementation. I think those can be removed now. Finally, I made a few comments inline, some of which repeat themselves in multiple places but the comment should apply to whole the file.

**Comment by VJalili** [comment]
@afgane Thanks for the comments, I updated the code accordingly. Few points: 

> Clearly, this is still AWS-only even if it implements the CloudBridge interface but I guess that's just the first step.

Correct. This PR sets the ground for User-based ObjectStore (initially discussed at #4314); which is a user-based and provider-agnostic ObjectStore. However, following the comments at #4314, I made `Cloud` with least possible changes to `S3ObjectStore` to (a) simplify comparison, (b) incremental and smaller impacts on the Galaxy codebase. Accordingly, you may see lots of `S3`-specific terminology and comments. I think we could be more provider-agnostic in future updates. 

I'll comment inline other comments.

File: lib/galaxy/dependencies/conditional-requirements.txt (hunk 2 above)
Comment (VJalili) [reply lib/galaxy/dependencies/conditional-requirements.txt]: No specific reason, I updated to the current latest version (also see starforge PR [#138](https://github.com/galaxyproject/starforge/pull/138)).

File: lib/galaxy/objectstore/cloud.py (hunk 3 above)
Comment (VJalili) [reply lib/galaxy/objectstore/cloud.py]: `boto` was only used to handle some exceptions. However, I replaced all `boto`-specific exceptions with a generic exception. This is totally a temporary patch till exceptions are properly wrapped and thrown by `CloudBridge`.

File: lib/galaxy/objectstore/cloud.py (hunk 4 above)
Comment (VJalili) [reply lib/galaxy/objectstore/cloud.py]: I agree, and as explained in the previous comments, as a temporary patch, replaced `S3`/`boto`-specific exceptions with generic exceptions.

File: lib/galaxy/objectstore/cloud.py (hunk 7 above)
Comment (VJalili) [reply lib/galaxy/objectstore/cloud.py]: Updated. But I'm postponing function signature change for future, just to keep the changes at this PR at a minimal possible level.

File: lib/galaxy/objectstore/cloud.py (hunk 9 above)
Comment (VJalili) [reply lib/galaxy/objectstore/cloud.py:461]: Agree. However, as I mentioned in other comments, I'm trying to keep `Cloud` very similar to `S3ObjectStore` (following the comments on the PR #4314). Using a general terminology might escalate the challenges :) But we'll surely use more general nomenclature in future updates.

**Comment by afgane** [comment]
It's generally not great to use the generic `Exception` if it can be helped but given this is an intermediary step, I think it's ok.

There are a couple of tests still failing though. How do the dependencies that are still getting built (i.e., the open PR for CloudBridge wheel) get pulled into the test env? Could that be it?

**Comment by VJalili** [comment]
Maybe the tests are failing because the current latest version of the `CloudBridge` is not built yet (https://github.com/galaxyproject/starforge/pull/138)

**Comment by VJalili** [comment]
@galaxybot test this

**Comment by nsoranzo** [comment]
[Unit tests](https://travis-ci.org/galaxyproject/galaxy/jobs/261703367) are failing with:
```
File "/home/travis/build/galaxyproject/galaxy/lib/galaxy/objectstore/cloud.py", line 25, in <module>

    from cloudbridge.cloud.factory import CloudProviderFactory, ProviderList

ImportError: No module named cloudbridge.cloud.factory
```

**Comment by bgruening** [comment]
@galaxybot test this

**Comment by VJalili** [comment]
@bgruening the test will fail because of a missing wheel (see [this PR](https://github.com/galaxyproject/starforge/pull/139)).

**Comment by VJalili** [comment]
@galaxybot test this

File: lib/galaxy/objectstore/cloud.py [hunk 10]
```diff
@@ -0,0 +1,565 @@
...
+                #     % (i, file_name, convert_bytes(f[2]), file_date, \
+                #     convert_bytes(deleted_amount), convert_bytes(delete_this_much)))
+            else:
+                log.debug("Cache cleaning done. Total space freed: %s", convert_bytes(deleted_amount))
+                return
+
+    def _get_bucket(self, bucket_name):
+        """ Sometimes a handle to a bucket is not established right away so try
+        it a few times. Raise error if connection is not established. """
+        for i in range(5):
```
Comment (nuwang) [inline lib/galaxy/objectstore/cloud.py:169]: @VJalili Can you elaborate on why this is needed and when it's triggered? Is there some parallel creation of a bucket going on? If not, I think this should be handled at the cloudbridge level, what do you think @afgane?

File: lib/galaxy/objectstore/cloud.py [hunk 11]
```diff
@@ -0,0 +1,565 @@
...
+            # Test if cache is large enough to hold the new file
+            if self.cache_size > 0 and key.size > self.cache_size:
+                log.critical("File %s is larger (%s) than the cache size (%s). Cannot download.",
+                             rel_path, key.size, self.cache_size)
+                return False
+            if self.use_axel:
+                log.debug("Parallel pulled key '%s' into cache to %s", rel_path, self._get_cache_path(rel_path))
+                ncores = multiprocessing.cpu_count()
+                url = key.generate_url(7200)
+                ret_code = subprocess.call("axel -a -n %s '%s'" % (ncores, url))
```
Comment (nuwang) [inline lib/galaxy/objectstore/cloud.py:299]: @VJalili @afgane This axel based downloading looks interesting - I wonder whether we should consider integrating similar functionality into cloudbridge? Not necessarily through axel, but at least a better implementation than the single threaded one we have now - maybe using an asyncio loop?

**Comment by afgane** [comment]
There is an issue with python 2.7 and asyncio but native download optimization would be nice.

null

**Comment by VJalili** [comment]
Some thing went awfully wrong with `rebase`. Please follow up #4487

**Comment by VJalili** [comment]
@nuwang @afgane due to an issue with this branch due to a white-space removal rebase, please follow up on a new PR #4487
